src/lkl/setup.c

```diff
@@ -1266,6 +1272,16 @@ static void* lkl_termination_thread(void* args)
         }
     }
 
+    SGXLKL_VERBOSE("calling lkl_sys_sync()\n");
+    ret = lkl_sys_sync();
```

Comment:
Have we done something here to prevent any threads that we failed to kill from doing system calls?  We can still end up with the disk in an undefined state if not:

 1. Sync call flushes everything
 2. User thread does some disk writes
 3. A few of them propagate to the VirtIO device
 4. We kill the kernel.

With Vikas' work, we now differentiate between userspace and kernel threads in lthreads.  Maybe we should set a flag so that our lthread scheduler refuses to run any userspace threads at this point?

Reply:
Right, but wouldn't we have to kill them in the LKL scheduler? If the lthread scheduler terminates a userspace thread while it holds the CPU lock, we would deadlock.

Reply:
Yes, I think we have three cases. Kernel threads are fine to run. Userspace threads executing in userspace are fine to stop. Userspace threads executing in the kernel are a problem. Hopefully LKL prevents any of the second category from becoming the third, so the only thing that can happen from running a userspace thread is that it exits the kernel.